build-tools/packages/build-cli/src/commands/generate/sourceEntrypoints.ts

```diff
@@ -0,0 +1,410 @@
+// eslint-disable-next-line import/no-internal-modules
+import { type ExportData, getExportPathFromPackage } from "../../library/packageExports.js";
+// AB#8118 tracks removing the barrel files and importing directly from the submodules, including disabling this rule.
+// eslint-disable-next-line import/no-internal-modules
+import { getApiExports, getPackageDocumentationText } from "../../library/typescriptApi.js";
+import type { CommandLogger } from "../../logging.js";
+
+/**
+ * Generates source entrypoints for Fluid Framework APIs to support API levels (/alpha, /beta. etc.).
+ */
+export default class GenerateSourceEntrypointsCommand extends BaseCommand<
+	typeof GenerateSourceEntrypointsCommand
+> {
+	static readonly description =
+		`Generates TypeScript source files that roll up APIs into different entrypoint files, defined by the "exports" field in package.json and organized by API tags`;
+
+	static readonly flags = {
+		mainEntrypoint: Flags.file({
+			description: "Main entrypoint file containing all untrimmed exports.",
+			default: "./src/index.ts",
+			exists: true,
+		}),
+		outDir: Flags.directory({
+			description: "Directory to emit entrypoint files.",
+			default: "./src/entrypoints/",
+			exists: true,
+		}),
+		...BaseCommand.flags,
+	};
+
+	public async run(): Promise<void> {
+		const { mainEntrypoint, outDir } = this.flags;
+
+		const packageJson = await readPackageJson();
+
+		// Read `rootDir` for tsconfig under `./src/entrypoints` or `outDir`
+		const { rootDir, tsconfigOutDir, emitDeclarationOnly } =
+			await getTsConfigCompilerOptions(outDir);
+
+		/**
+		 * Example of `mapApiTagToExportPath` for a package where `emitDeclarationOnly` is false and it's an ESM package
+		 * [[beta, `{ relPath: "./lib/entrypoints/beta.js", conditions: [], isTypeOnly: false }`],
+		 * [public, `{ relPath: "./lib/entrypoints/public.js", conditions: [], isTypeOnly: false }`]]
+		 */
+		const mapApiTagToExportPath: Map<ApiTag, ExportData> = mapExportPathToApiTag(
+			packageJson,
+			emitDeclarationOnly,
+			this.logger,
+		);
+
+		/**
+		 * Example of `mapApiTagToSourcePath` for a package where rootDir is ./src/entrypoints/
+		 * [["./src/entrypoints/beta.ts", "beta"],
+		 * ["./src/entrypoints/alpha.ts", "alpha"],
+		 * ["./src/entrypoints/public.ts", "public"],
+		 * ["./src/entrypoints/legacy.ts", "legacy"]]
+		 */
+		const mapApiTagToSourcePath: Map<string, ApiTag> = await mapSourcePathToApiTag(rootDir);
+
+		/**
+		 * Example of `mapSourceToExportPath` for a package
+		 * [[beta, `{ relPath: "./src/entrypoints/beta.js", conditions: [], isTypeOnly: false }`],
+		 * [public, `{ relPath: "./src/entrypoints/public.js", conditions: [], isTypeOnly: false }`]]
+		 */
+		const mapSourceToExportPath: Map<ApiTag, ExportData> = getOutputConfiguration(
+			mapApiTagToSourcePath,
+			mapApiTagToExportPath,
+			rootDir,
+			tsconfigOutDir,
+			this.logger,
+		);
```

Comment:
Seems overcomplicated now. Can't we just take `mapApiTagToExportPath` that has API tag key and value with relPath of .js/.d.ts output and remap the relPath to the source?
Apply:
relPath.replace(tsconfigOutDir, tsconfigRootDir).replace(/\.js$|\.d\.ts$/, ".ts");
and be done?